Add support for Cloudflare R2 cloud storage

hlky opened the pull request:
## Description of changes:

Add support for Cloudflare R2 cloud storage

```
>>> import numpy as np
>>> from PIL import Image
>>> from streaming import MDSWriter
>>>
>>> # Local or remote directory in which to store the compressed output files
>>> data_dir = 'r2://test'
>>>
>>> # A dictionary mapping input fields to their data types
>>> columns = {
...     'image': 'jpeg',
...     'class': 'int'
... }
>>>
>>> # Shard compression, if any
>>> compression = 'zstd'
>>>
>>> # Save the samples as shards using MDSWriter
>>> with MDSWriter(out=data_dir, columns=columns, compression=compression) as out:
...     for i in range(10000):
...         sample = {
...             'image': Image.fromarray(np.random.randint(0, 256, (32, 32, 3), np.uint8)),
...             'class': np.random.randint(10),
...         }
...         out.write(sample)
...
>>> from torch.utils.data import DataLoader
>>> from streaming import StreamingDataset
>>>
>>> # Remote path where full dataset is persistently stored
>>> remote = 'r2://test'
>>>
>>> # Local working dir where dataset is cached during operation
>>> local = './test'
>>>
>>> # Create streaming dataset
>>> dataset = StreamingDataset(local=local, remote=remote, shuffle=True)
>>>
>>> # Let's see what is in sample #1337...
>>> sample = dataset[1337]
>>> img = sample['image']
>>> cls = sample['class']
>>>
>>> # Create PyTorch DataLoader
>>> dataloader = DataLoader(dataset)
>>> sample
{'class': 6, 'image': <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=32x32 at 0x23A3B83BFA0>}
>>>
```

![image](https://github.com/mosaicml/streaming/assets/106811348/d1277f57-ba6c-4196-a320-144221b24d87)


## Issue #, if available:


## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [ ] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [ ] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

<!--
Thanks so much for contributing to Streaming! We really appreciate it :)
-->

karan6181 commented:
Thanks @hlky for creating a PR.

hlky commented:
Let me know if there any changes required, thanks.

karan6181 commented:
Overall, the PR looks great @hlky . Thank you for your contribution. I have some minor comments to it.

Cloudflare r2 is an [s3-compatible](https://developers.cloudflare.com/r2/api/s3/api/) storage system. Users credentials can be set explicitly or it can retrieved implicitly  from `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` , here is the reference [link](https://developers.cloudflare.com/r2/examples/aws/boto3/) for the same. Hence, instead of having a separate environment variables for R2 such as `R2_ACCESS_KEY_ID` and `R2_SECRET_ACCESS_KEY`. Instead, we can rely on the default AWS environment variable names such as `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`. The benefit of this is that it can work in both the cases as mentioned [here](https://developers.cloudflare.com/r2/examples/aws/boto3/) where user just have to set the credentials once as part of `~/.aws/credentials` and `~/.aws/config`.

**Here is my proposal:**

- You can remove `R2_ACCESS_KEY_ID` and `R2_SECRET_ACCESS_KEY` and replace `R2_ENDPOINT_URL` with `S3_ENDPOINT_URL`.
- you can update the R2 documentation by pointing it [here](https://docs.mosaicml.com/projects/streaming/en/stable/how_to_guides/configure_cloud_storage_cred.html#others) for the credentials setup. There will be an additional step required to set the flag `S3_ENDPOINT_URL`. Also mentioned this as part of R2 documentation.
- On the streaming dataset side, streaming code will just have to do

```python
r2_client = session.client('s3', region_name='auto', endpoint_url=os.environ['S3_ENDPOINT_URL'])
```

- For the multiple object stores, for example, if user would like to keep an AWS S3 credentials as well as R2 credentials part of `~/.aws/credentials` and `~/.aws/config`, then they can do so by separating the credentials in a different `AWS_PROFILE`. For example, user just need to set all of credentials under `AWS_PROFILE` section and set it either as `os.environ['AWS_PROFILE'] = '<profile_name>'` or `export AWS_PROFILE='<profile_name>'`. You can also add this point as part of R2 documentation. The `~/.aws/credentials `  would look something like this as shown below. In this case, `os.environ['AWS_PROFILE'] = 'r2'`.

```bash
[default]
aws_access_key_id=<your_aws_access_key>
aws_secret_access_key=<your_aws_secret_key>
[r2]
aws_access_key_id=<your_cloudflare_access_key>
aws_secret_access_key=<your_cloudflare_secret_key>
```

The `~/.aws/config` would look something like this as shown below:

```bash
[default]
region=us-west-2
[r2]
region=us-west-2
```

- On the streaming side, there is no change in the API, it will still have to do the same thing since boto3 imports `AWS_PROFILE` during runtime.

```python
r2_client = session.client('s3', region_name='auto', endpoint_url=os.environ['S3_ENDPOINT_URL'])
```

Please let us know if you have any questions. Thank you!

karan6181 commented on streaming/base/storage/download.py:
```diff
@@ -198,6 +198,37 @@ def download_from_oci(remote: str, local: str) -> None:
+    """Download a file from remote Cloudflare R2 to local.
+
+    Args:
+        remote (str): Remote path (S3).
```
nit:`Remote path (S3).` -> `Remote path (R2).`

karan6181 replied:
Addressed as part of https://github.com/mosaicml/streaming/pull/255/commits/3b7e2068f524d6ff2a3b9de4908ff7064a8eb5eb.

karan6181 commented on streaming/base/storage/download.py:
```diff
@@ -223,6 +254,10 @@ def download_file(remote: Optional[str], local: str, timeout: float):
     if os.path.exists(local):
         return
 
+    local = local.replace('\\', '/')
```
Is this for windows file system ?

hlky replied:
Yes, os.path.join returns paths with `\` on Windows and remote file name e.g. `s3://test\example` doesn't work

karan6181 commented on streaming/base/storage/upload.py:
```diff
@@ -393,6 +396,89 @@ def check_bucket_exists(self, remote: str):
+        """Raise an exception if the bucket does not exist.
+
+        Args:
+            remote (str): S3 bucket path.
```
nit: S3 bucket path. -> `r2 bucket path.`

karan6181 replied:
Addressed as part of https://github.com/mosaicml/streaming/pull/255/commits/3b7e2068f524d6ff2a3b9de4908ff7064a8eb5eb.

karan6181 commented on tests/conftest.py:
```diff
@@ -17,6 +17,7 @@
 MY_BUCKET = 'streaming-test-bucket'
 MY_PREFIX = 'train'
 GCS_URL = 'https://storage.googleapis.com'
+R2_URL = 'https://c4c9097c86eeef198b16fd088f5f99d6.r2.cloudflarestorage.com'
```
Is this a real account id? Can we use any other alphanumeric value such as `abcdefghijkl1234567` ? since we are anyhow mocking the response!!

hlky replied:
Random value gives SSL error and test fails, I've replaced it with just `'https://r2.cloudflarestorage.com'`

karan6181 replied:
Addressed as part of https://github.com/mosaicml/streaming/pull/255/commits/3b7e2068f524d6ff2a3b9de4908ff7064a8eb5eb.

karan6181 commented on streaming/base/storage/download.py:
```diff
@@ -144,7 +144,7 @@ def download_from_gcs(remote: str, local: str) -> None:
 
     Args:
-        remote (str): Remote path (S3).
+        remote (str): Remote path (GCS).
```
Thank You!

karan6181 reviewed (commented):
Thank you for addressing the comments. Do you have a download and upload shard results from R2 ? It would be great if you can add that in the PR description.

hlky commented:
As suggested I have removed R2 specific environment variables, replaced `R2_ENDPOINT_URL` with `S3_ENDPOINT_URL`, and updated documentation.

There is a fix for paths on Windows in R2Uploader's `upload_file` too, this could be replicated to other upload classes. Tests on Windows are also spotty with many tests failing from `Invalid Cloud provider prefix.` and others
```
Expected: download_from_gcs('gs://streaming-test-bucket\\train\\file.txt', 'C:\\Users\\user\\AppData\\Local\\Temp\\tmp16qu_of_\\file.txt')
Actual: download_from_gcs('gs://streaming-test-bucket/train/file.txt', 'C:/Users/user/AppData/Local/Temp/tmp16qu_of_/file.txt')
```
```
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid bucket name "streaming-test-bucket\train\file.txt": Bucket name must match the regex "^[a-zA-Z0-9.\-_]{1,255}$" or be an ARN matching the regex "^arn:(aws).*:(s3|s3-object-lambda):[a-z\-0-9]*:[0-9]{12}:accesspoint[/:][a-zA-Z0-9\-.]{1,63}$|^arn:(aws).*:s3-outposts:[a-z\-0-9]+:[0-9]{12}:outpost[/:][a-zA-Z0-9\-]{1,63}[/:]accesspoint[/:][a-zA-Z0-9\-]{1,63}$"
```

I have also updated the description.

karan6181 reviewed (approved):
LGTM. Thank You!

bandish-shah commented:
@hlky thanks for doing this, it's great to see the OSS community get involved!  Please let us know what else we can add or improve.

hlky commented:
No problem, happy to help, Streaming seems to be a great alternative to WebDataset. I'll check further into Windows support as this would be useful for me.
